File: simulator/Organisms/Animal.py
from simulator.Organism import Organism
import random


class Animal(Organism):
    def __init__(self, current_world, x_coordinate, y_coordinate, strength, initiative, is_alive, token, image):
        super().__init__(current_world, x_coordinate, y_coordinate, strength, initiative, is_alive, token, image)

    @staticmethod
    def is_animal():
        return True

    def action(self):
        if self.get_age() != 0:
            self.move()

    def collision(self, predator):
        text = ""
        if predator.get_token() != self.get_token():
            text += str(predator.get_token()) + " attacks " + str(self.get_token()) + " "
            if (self.get_strength() > predator.get_strength()
                    or (self.get_strength() == predator.get_strength() and self.get_age() > predator.get_age())):
                text += str(self.get_token()) + " wins"
                self.get_world().add_log(text)
                self.get_world().del_organism(predator)
            else:
                text += str(predator.get_token()) + " wins"
                self.get_world().add_log(text)
                tmp_x = self.get_x()
                tmp_y = self.get_y()
                self.get_world().del_organism(self)
                predator.new_location(tmp_x, tmp_y)
        elif self.get_age() > 3 and predator.get_age() > 3:
            self.get_world().new_organism(self)

    def move(self):
        legal_move = False
        while not legal_move:
            x = 0
            y = 0
            direction = random.randint(0, 3)
            if direction == 0:
                y += self.get_move_dist()
                if not self.get_world().is_outside(self.get_x() + x, self.get_y() + y):
                    legal_move = True
            elif direction == 1:
                x += self.get_move_dist()
                if not self.get_world().is_outside(self.get_x() + x, self.get_y() + y):
                    legal_move = True
            elif direction == 2:
                y -= self.get_move_dist()
                if not self.get_world().is_outside(self.get_x() + x, self.get_y() + y):
                    legal_move = True
            elif direction == 3:
                x -= self.get_move_dist()
                if not self.get_world().is_outside(self.get_x() + x, self.get_y() + y):
                    legal_move = True
            if legal_move:
                if self.get_world().is_occupied(self.get_x() + x, self.get_y() + y):
                    self.get_world().get_field(self.get_x() + x, self.get_y() + y).collision(self)
                else:
                    new_x = self.get_x() + x
                    new_y = self.get_y() + y
                    self.new_location(new_x, new_y)

File: world_simulator_cpp/Animal.cpp
#include <iostream>
#include "World.h"
#include "Animal.h"
#include "enumList.h"
using namespace std;

Animal::Animal(World* currentWorld, int X_coordinate, int Y_coordinate, int strength, int initiative, char token)
	:Organism(currentWorld, X_coordinate, Y_coordinate, strength, initiative, token) {
}

bool Animal::IsAnimal() {
	return true;
}

void Animal::Action() {
	if (this->GetAge() == 0)
		return;
	Move();
}
void Animal::Collision(Organism* predator, Organism* prey) {
	if (predator->GetToken() != prey->GetToken()) {
		cout << predator->GetToken() << " attacks " << prey->GetToken() << " on x: " << prey->GetX() << " y: " << prey->GetY() << endl;
		if (prey->GetStrength() > predator->GetStrength()) {
			this->GetWorld()->DelFromWorldMap(predator);
			this->GetWorld()->DelFromOrganismList(predator);
			delete predator;
		}
		else {
			predator->ChangeLocation(prey->GetX(), prey->GetY());
			this->GetWorld()->DelFromWorldMap(prey);
			this->GetWorld()->DelFromOrganismList(prey);
			predator->Draw();
		}
	}
	else {
		if (predator->GetAge() > 3 && prey->GetAge() > 3)
			this->currentWorld->GiveBirth(predator);
	}
}
void Animal::Move() {
	bool LegalMove = false;
	int x, y;
	while (!LegalMove) {
		x = 0;
		y = 0;
		int direction = rand() % 4;
		if (direction == DOWN) {
			y++;
			if (!currentWorld->IsOutsideBorder(x + GetX(), y + GetY()))
				LegalMove = true;
		}
		else if (direction == RIGHT) {
			x++;
			if (!currentWorld->IsOutsideBorder(x + GetX(), y + GetY()))
				LegalMove = true;
		}
		else if (direction == UP) {
			y--;
			if (!currentWorld->IsOutsideBorder(x + GetX(), y + GetY()))
				LegalMove = true;
		}
		else {
			x--;
			if (!currentWorld->IsOutsideBorder(x + GetX(), y + GetY()))
				LegalMove = true;
		}
		if (LegalMove) {
			if (this->currentWorld->IsFieldOccupied(x + this->GetX(), y + this->GetY())){
				this->currentWorld->GetField(x + this->GetX(), y + this->GetY())->Collision(this, this->currentWorld->GetField(x + this->GetX(), y + this->GetY()));
			}
			else {
				ChangeLocation(x + GetX(), y + GetY());
				cout << this->GetToken() << " moves to x: " << this->GetX() << " y: " << this->GetY() << endl;
			}	
		}
	}

}

Animal::~Animal() {

}
